debugging.md

```diff
@@ -4,73 +4,127 @@ This page details information on common problems you might have with a command,
 * Use a plain plain text editor (Notepad, Notepad++, Sublime, Code), **not** a word processor or rich text editor (Microsoft Word, Wordpad, Textedit)
 * Make sure your file's encoding is UTF-8 (without BOM) - this is not the default in many programs! ([Notepad](http://i.imgur.com/R4yFjAQ.png), [Notepad++](http://i.imgur.com/8AsDJ3F.png), [Sublime](http://i.imgur.com/63rsYOB.png), [Code](http://i.imgur.com/dmOqy0y.png))
-* The namespace folder is not optional, functions should **not** be directly inside `data/functions/`
+* The namespace folder is not optional, functions should **not** be directly inside `data/functions/`, they must be in `data/<namespace>/functions/`
+* Don't forget to save changes if editing directly the datapack (with programs like VScode), normaly `ctrl+s` (Windows/Lunix) or `cmd+s` (Mac)
 * Don't forget to use `/reload` to reload the functions after making changes
+  * Make sure you have the datapack enabled. You can enable the datapack by typing `/datapack enable "<datapack name>"`
+  * When edition world generation you need to leave and rejoin the world in order to save changes (or restart the server).
 * Check that you're saving to the place you think you're saving to (right world, right namespace), and running the function you intend to
-* Recursive/looping functions will run `maxCommandChainLength` commands in one tick, then stop
+* Recursive/looping functions will run `maxCommandChainLength` commands in one tick, then stop, the default value of this gamerule is `65536`
+  * To check if a function did **not** end (so it has been cut off by this gamerule) you can set a scoreboard value of a [fakeplayer](wiki/questions/fakeplayer) to 1, run all commands and set it to 0 after that (that command should be the last one in the whole tick). In the next tick you can check if the value is 1 or 0, if its 1, the function has been cutted.
+* Make sure the tick/load function tag is specifing the correct function
+* Functions can run any command that requires a higher level permission than 2.
+  * You can override this if you set `function-level-permission` in `server.propieties` to a higher number
+* If a function has a syntax error, it will **not** show up in the autocomplete and it will flail, if you try to run it
+  * All commands must have a correct syntax in order to run the function
+  * You can start the function with a `/say` command, so you will know if the function has been executed
+* In snapshot [24w21b](https://www.minecraft.net/en-us/article/minecraft-snapshot-24w21a) (for 1.21) some registry types that used legacy datapack directory names (based on plural name of element) have been renamed to match registry name. Affected directories:
+  * `structures` -> `structure`
+  * `advancements` -> `advancement`
+  * `recipes` -> `recipe`
+  * `loot_tables` -> `loot_table`
+  * `predicates` -> `predicate`
+  * `item_modifiers` -> `item_modifier`
+  * `functions` -> `function`
+  * `tags/functions` -> `tags/function`
```

Comment:
Not 24w21b, but 24w21a (not so important), but it is not the only snapshot in which directories were renamed. In snapshot [24w19a](https://www.minecraft.net/en-us/article/minecraft-snapshot-24w19a), tag directories were also renamed.

> tags/items -> tags/item
> tags/blocks -> tags/block
> tags/entity_types -> tags/entity_type
> tags/fluids -> tags/fluid
> tags/game_events -> tags/game_event